File: 0015-3sum/0015-3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(),nums.end());
        vector<vector<int>>res;
        int n = nums.size();
        for(int i=0;i<n-2;i++)
        {
            int l = i+1;
            int r = n-1;
            if(nums[i] > 0)
            {
                break;
            }
            if(i > 0 && nums[i] == nums[i-1])
            {
                continue;
            }
            while(l < r)
            {
                int sm = nums[i]+nums[l]+nums[r];
                if(sm > 0)
                {
                    r--;
                }
                else if(sm < 0)
                {
                    l++;
                }
                else
                {
                    vector<int>t = {nums[i],nums[l],nums[r]};
                    res.push_back(t);
                    while(l < r && nums[l] == nums[l+1])
                    {
                        l++;
                    }
                    while(l < r && nums[r] == nums[r-1])
                    {
                        r--;
                    }
                    l++;
                    r--;
                }
            }
        }
        return res;
    }
};

File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: list[int]) -> list[list[int]]:
        nums.sort()
        n = len(nums)
        res = []

        for i in range(n-2):
            if i > 0 and nums[i] == nums[i-1]:
                continue  # skip duplicate `i`
            j, k = i + 1, n - 1
            while j < k:
                total = nums[i] + nums[j] + nums[k]
                if total == 0:
                    res.append([nums[i], nums[j], nums[k]])
                    j += 1
                    k -= 1
                    # skip duplicates
                    while j < k and nums[j] == nums[j-1]:
                        j += 1
                    while j < k and nums[k] == nums[k+1]:
                        k -= 1
                elif total < 0:
                    j += 1
                else:
                    k -= 1
        return res